### PR TITLE
parse derive input with darling

### PR DESCRIPTION
**Pull request by DCjanus**
with `darling`, code could be simple and easy to read.

### CODE REVIEW

**Comment by djc** [comment]
What are you trying to achieve here?

**Comment by DCjanus** [comment]
> What are you trying to achieve here?
Make the code shorter and simple.

File: askama_derive/src/input.rs [hunk 1]
```diff
@@ -160,25 +102,20 @@ impl<'a> TemplateInput<'a> {
...
-            }
-        }
-
-        let escaper = escaper.unwrap_or_else(|| {
-            panic!("no escaper defined for extension '{}'", extension);
-        });
+        let escaper = config
+            .escapers
+            .iter()
+            .find(|(extensions, _)| extensions.contains(&extension))
```
Comment (djc) [inline askama_derive/src/input.rs]: Let's use `find_map()` here?

File: askama_derive/src/input.rs [hunk 2]
```diff
@@ -25,106 +37,36 @@ impl<'a> TemplateInput<'a> {
...
-        // Validate the `source` and `ext` value together, since they are
-        // related. In case `source` was used instead of `path`, the value
-        // of `ext` is merged into a synthetic `path` value here.
-        let source = source.expect("template path or source not found in attributes");
+        // Since 'source' and 'path' are related. In case `source` was used instead
+        // of `path`, the value of `ext` is merged into a synthetic `path` value here.
+        let source = source
+            .map(Source::Source)
+            .unwrap_or_else(|| Source::Path(path.unwrap()));
         let path = match (&source, &ext) {
```
Comment (djc) [inline askama_derive/src/input.rs]: If you're going to do this, let's also move this login into a `Source::path()` method.

File: askama_derive/src/input.rs [hunk 3]
```diff
@@ -25,106 +37,36 @@ impl<'a> TemplateInput<'a> {
...
-                panic!(
-                    "unsupported attribute key '{}' found",
-                    pair.path.to_token_stream()
-                )
-            }
-        }
+        assert_ne!(
+            source.is_some(),
+            path.is_some(),
+            "One of path or source must exist and they are mutually exclusive",
```
Comment (djc) [inline askama_derive/src/input.rs]: Let's lowercase the `One` here to be consistent, and simplify the "and they are mutually exclusive to " (but not both)"?

File: askama_derive/src/input.rs [hunk 4]
```diff
@@ -17,6 +18,17 @@ pub struct TemplateInput<'a> {
     pub path: PathBuf,
 }
 
+#[derive(FromDeriveInput, Default)]
+#[darling(attributes(template), default)]
+struct TemplateInputParser {
```
Comment (djc) [inline askama_derive/src/input.rs]: Let's just call this `TemplateAttributes`.

File: askama_derive/src/input.rs [hunk 5]
```diff
@@ -25,106 +37,36 @@ impl<'a> TemplateInput<'a> {
...
-            _ => panic!("attribute 'template' has incorrect type"),
-        };
+        let TemplateInputParser {
+            path,
+            source,
+            print,
+            escape,
+            ext,
+            syntax,
+        } = FromDeriveInput::from_derive_input(ast)
```
Comment (djc) [inline askama_derive/src/input.rs]: And then call this as `TemplateAttributes::from_derive_input(ast)`?

**Comment by DCjanus** [comment]
Sorry for my mistake, I close this PR temporary and try to clean my code up when I  time. Thanks for your help.

**Comment by djc** [comment]
Sorry to be a bit harsh in my response. These are good changes that I'd like to merge, I just get defensive when receiving refactoring-only PRs with little context out of the blue. So next time, just writing up an issue first would help, or just giving a little bit of context in the PR description or commit message would also be great. Thanks for your contributions to askama!

**Comment by DCjanus** [comment]
Sorry for delay. I just find that my words sounds like I'm complaining about something.

I'm not a native speaker, and it's really hard to show my real emotion in English for me, and somethins, Google translate make me looks like a jerk :) 

BTW, 'harsh' is a good attitude for an open source project.

It's the Chinese New Year in a few days, and I have a lot of things to deal with, so I don’t have time to continue this PR for a while. This is why I temporarily closed this PR.

This PR has been on my TODO list, I will finish this one day.

**Comment by djc** [comment]
No problem at all. Enjoy the New Year's events!

**Pull request closed without merging**